chore: optimize example docker files using multi-stage builds

sapkota-aayush opened the pull request:
Optimize example Dockerfiles - Reduce image size by 60%

Fixed broken multi-stage builds in all 17 example Dockerfiles by changing FROM builder AS udf to FROM python:3.10-slim-bullseye AS udf. This removes build tools (gcc, git, build-essential, Poetry) from final images and only copies the virtual environment and application code.
Results
Before: ~971MB images (with build tools)
After: ~385MB images (clean runtime only)
Improvement: 60% size reduction (586MB saved per image)
Testing
Verified with examples/reduce/counter/Dockerfile - build tools successfully removed, functionality preserved.
Files Changed
All 17 example Dockerfiles across map, reduce, sink, source, and sourcetransform categories.
Closes #181

vigith commented:
The jobs are failing with the following error, please run `poetry lock`

```
Using virtualenv: /home/runner/.cache/pypoetry/virtualenvs/pynumaflow-Fn3toN8H-py3.9
Installing dependencies from lock file

pyproject.toml changed significantly since poetry.lock was last generated. Run `poetry lock` to fix the lock file.
Error: Process completed with exit code 1.
```

sapkota-aayush commented on pyproject.toml:
```diff
@@ -31,6 +31,7 @@ protobuf = ">=3.20,<6.0"
 aiorun = "^2023.7"
 uvloop = "^0.19.0"
 psutil = "^6.0.0"
+numpy = "^1.26.0"
```
> why would you want to add `numpy`?

Thank you for catching that! I had temporarily added numpy while testing one of the examples, but it’s not actually required for any of the examples or the project itself. I’ve now removed it from pyproject.toml and updated the lock file accordingly.

sapkota-aayush commented:
> @sapkota-aayush You can use the inbuild `make lint` command for the repo
@vigith  @kohlisid 

I wasn’t able to run make lint directly because of some setup issues, but I manually fixed the formatting errors with Black and double-checked everything looks good now. That should take care of the lint errors.

Whenever you get a chance, maybe you could suggest the best way to run make lint on my setup?

Thanks!

vigith commented:
looks like it ran successfully now.

kohlisid commented:
You can try to setup a virtual env for the project, and install the required dependencies from the project toml file

sapkota-aayush commented:
Thanks @vigith  Appreciate the tip @kohlisid — I’ll try setting up the virtual environment properly for smoother runs next time.

Also noticed that even after optimizing the Dockerfile size, each UDF still takes over 2 mins to build, which feels a bit much. Not sure if there’s a better way — maybe we could try caching or something else to speed it up?

Open to any suggestions — just trying to streamline things a bit. Appreciate the help as always!

vigith commented:
did you look into https://github.com/numaproj/numaflow-python/issues/181#issuecomment-2224070082 ?

kohlisid commented:
@sapkota-aayush We can try to divide it into three multi stage layers
1. Base
2. Env setup
3. Builder 

Then for each we can look at specific optimizations

sapkota-aayush commented:
@vigith @kohlisid 
I’ve optimized all the example Dockerfiles using a 3-layer multi-stage build, fixed some file and formatting issues, and removed unnecessary dependencies and tested them as well. The only one I didn’t fully update is examples/reduce/batchmap/flatmap/Dockerfile since it’s missing a pyproject.toml and looks incomplete. If you could run the tests and let me know if anything else needs fixing or if you have any feedback, that’d be great. Thanks!

vigith commented on pyproject.toml:
```diff
@@ -80,7 +80,6 @@ extend-exclude = [
 ]
-output-format = "full"
-[tool.ruff.lint]
+format = "grouped"
```
you can upgrade the ruff version we use. I am not an expert in these tools either :)

kohlisid replied:
https://github.com/numaproj/numaflow-python/blob/main/pyproject.toml#L38

vigith commented on pyproject.toml:
```diff
@@ -29,7 +29,7 @@ google-api-core = "^2.11.0"
 protobuf = ">=3.20,<6.0"
 aiorun = "^2023.7"
-uvloop = "^0.19.0"
+uvloop = { version = "^0.19.0", markers = "sys_platform == 'linux'" }
```
please revert this?

sapkota-aayush replied:
Done.

kohlisid commented on DOCKER_OPTIMIZATION.md:
```diff
@@ -0,0 +1,229 @@
+# Docker Build Optimization for NumaFlow Python UDFs
```
@sapkota-aayush 
Could you move this to a docs/ directory?

sapkota-aayush replied:
Just to confirm, since I’m currently working only in the numaflow-python repository (which doesn’t have a docs/ directory by default), would you like me to create a docs/ directory here and move DOCKER_OPTIMIZATION.md into it? Or should this documentation be added to the main numaflow project’s docs/ directory instead?

kohlisid replied:
Yes you can create a docs directory for this repo

sapkota-aayush replied:
> Yes you can create a docs directory for this repo

Done

kohlisid commented on pyproject.toml:
```diff
@@ -80,7 +80,5 @@ extend-exclude = [
     "*_pb2*.py",
     "*.pyi"
 ]
-output-format = "full"
```
What we the reason to change this?

sapkota-aayush replied:
> What was the reason to change this?

Hi @kohlisid,
Thanks for pointing that out! I’m still getting the hang of these tools and configurations, as a lot of this is new to me. I initially added output-format = "full" thinking it would help, but realized our current Ruff version doesn’t support it, which caused the errors.
I’ve removed it now so everything matches the main branch and works with our current setup.

kohlisid commented:
@sapkota-aayush Have you tested the new images end-to-end by running in a pipeline?

sapkota-aayush commented:
> @sapkota-aayush Have you tested the new images end-to-end by running in a pipeline?

Yes, I have tested the new images end-to-end by running them in a Numaflow pipeline on Kubernetes.
The pipeline was deployed successfully, all pods were running and healthy, and the custom Docker image was used in the workflow.
Please see the attached screenshot for details.
Let me know if you need any more information!
![test3](https://github.com/user-attachments/assets/2283bf0b-2427-4c0f-8cc0-0f265aed804e)
![test2](https://github.com/user-attachments/assets/d5729379-34db-4112-9dad-2f92fa720c4a)
![test1](https://github.com/user-attachments/assets/65fb7262-1db8-4c19-833d-3474300004e4)

sapkota-aayush commented:
> Thanks for taking it up and the collaboration on this @sapkota-aayush :D

Thank you @vigith and @kohlisid for your guidance. Learned a lot.
